dhis2/capture-app
# feat: [DHIS2-15475] Form Field Plugins

This PR adds the experimental support of Form Field Plugins.

How to test this feature:

### Hosted version

I have uploaded and configured everything that you need to test this out in a separate plugins instance. It will be used by many other things as well, but for now, it's mostly these PR's.
Find the `Capture - Plugins` app and inspect as you will.

Instance: https://dev.im.dhis2.org/tracker-plugins

Go directly to [plugin form](https://dev.im.dhis2.org/tracker-plugins/api/apps/capture-plugins/index.html#/new?orgUnitId=DiszpKrYNg8&programId=IpHINAT79UW)

### Test locally

There are some small steps that you need to do before testing this PR out.

- Clone the tracker-plugin repo on my GH-account and run it on port 3001 (plugin will be at 3002)
    - You can also build the plugin and upload it through `App management` (Good if you get CORS-errors)
- Set up the plugin configuration in DataStore Management
    - Namespace: `capture`
    - Key: `dataEntryForms`
    - Input your form configuration (see `Form Configuration` for a detailed breakdown)
-  Open the capture app

### Form configuration (example):

The form configuration array defines the way your form looks and functions. Each object in this array defines one section, with each section containing a set of Tracked Entity Attributes or Plugins. This configuration will display the Civil Registry Plugin in the Child Programme context.

```json
{
  "IpHINAT79UW": [
    {
      "id": "test",
      "name": "Section 1",
      "elements": [
        {
          "id": "w75KJ2mc4zz",
          "type": "TrackedEntityAttribute"
        },
        {
          "id": "zDhUuAYrxNC",
          "type": "TrackedEntityAttribute"
        }
      ]
    },
    {
      "id": "plugin",
      "name": "Section 2",
      "elements": [
        {
          "id": "plugin",
          "name": "civil_registry",
          "pluginSource": "http://localhost:3002/plugin.html",
          "type": "plugin",
          "fieldMap": [
            {
              "IdFromApp": "w75KJ2mc4zz",
              "IdFromPlugin": "firstName",
              "objectType": "TrackedEntityAttribute"
            },
            {
              "IdFromApp": "zDhUuAYrxNC",
              "IdFromPlugin": "lastName",
              "objectType": "TrackedEntityAttribute"
            },
            {
              "IdFromApp": "yx2MCEAfU8N",
              "IdFromPlugin": "TestAttribute",
              "objectType": "Attribute"
            },
            {
              "IdFromApp": "cejWyOfXge6",
              "IdFromPlugin": "gender",
              "objectType": "TrackedEntityAttribute"
            }
          ]
        }
      ]
    }
  ]
}
```
